The result of format_annotation() has been changed with python 3.7.0a4+

### Problem
- The result of format_annotation() has been changed with python 3.7.0a4+.

#### Procedure to reproduce the problem
- Run tests of Sphinx with python 3.7.0a4+

#### Error logs / results
https://travis-ci.org/sphinx-doc/sphinx/jobs/332762559#L2003

Since py370a4+, the implementation of `typing` module has been changed (ref: https://github.com/python/cpython/pull/4906)
```
$ python3.6
Python 3.6.4 (default, Dec 20 2017, 01:09:24)
[GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import typing
>>> isinstance(typing.List, type)
True
>>> typing.GenericMeta
<class 'typing.GenericMeta'>
>>> typing.CallableMeta
<class 'typing.CallableMeta'>
```
```
$ python3.7
Python 3.7.0a4+ (heads/master:131fd7f, Jan 24 2018, 23:03:18)
[Clang 8.0.0 (clang-800.0.42.1)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import typing
>>> isinstance(typing.List, type)
False
>>> typing.GenericMeta
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'typing' has no attribute 'GenericMeta'
>>> typing.CallableMeta
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'typing' has no attribute 'CallableMeta'
```

For this reason, the result of `format_annotation()` has been changed.
```
>       assert formatargspec(f, *getargspec(f)) == expected
E       AssertionError: assert '(x: typing.L...st[-T_contra]' == '(x: typing.Li...ist[T_contra]'
E         - (x: typing.List[~T], y: typing.List[+T_co], z: T) -> typing.List[-T_contra]
E         ?                 -                   -                            -
E         + (x: typing.List[T], y: typing.List[T_co], z: T) -> typing.List[T_contra]
```
There are `+`, `-` and `~` prefix on annotation.

--- a/CHANGES
+++ b/CHANGES
@@ -24,6 +24,7 @@ Bugs fixed
 * #4415: autodoc classifies inherited classmethods as regular methods
 * #4415: autodoc classifies inherited staticmethods as regular methods
 * #4472: DOCUMENTATION_OPTIONS is not defined
+* #4490: autodoc: type annotation is broken with python 3.7.0a4+
 
 Testing
 --------

--- a/sphinx/util/inspect.py
+++ b/sphinx/util/inspect.py
@@ -431,6 +431,55 @@ class Signature(object):
 
         Displaying complex types from ``typing`` relies on its private API.
         """
+        if sys.version_info >= (3, 7):  # py37+
+            return self.format_annotation_new(annotation)
+        else:
+            return self.format_annotation_old(annotation)
+
+    def format_annotation_new(self, annotation):
+        # type: (Any) -> str
+        """format_annotation() for py37+"""
+        module = getattr(annotation, '__module__', None)
+        if isinstance(annotation, string_types):
+            return annotation  # type: ignore
+        elif isinstance(annotation, typing.TypeVar):  # type: ignore
+            return annotation.__name__
+        elif not annotation:
+            return repr(annotation)
+        elif module == 'builtins':
+            return annotation.__qualname__
+        elif annotation is Ellipsis:
+            return '...'
+
+        if module == 'typing':
+            if getattr(annotation, '_name', None):
+                qualname = annotation._name
+            elif getattr(annotation, '__qualname__', None):
+                qualname = annotation.__qualname__
+            else:
+                qualname = self.format_annotation(annotation.__origin__)  # ex. Union
+        elif hasattr(annotation, '__qualname__'):
+            qualname = '%s.%s' % (module, annotation.__qualname__)
+        else:
+            qualname = repr(annotation)
+
+        if getattr(annotation, '__args__', None):
+            if qualname == 'Union':
+                args = ', '.join(self.format_annotation(a) for a in annotation.__args__)
+                return '%s[%s]' % (qualname, args)
+            elif qualname == 'Callable':
+                args = ', '.join(self.format_annotation(a) for a in annotation.__args__[:-1])
+                returns = self.format_annotation(annotation.__args__[-1])
+                return '%s[[%s], %s]' % (qualname, args, returns)
+            else:
+                args = ', '.join(self.format_annotation(a) for a in annotation.__args__)
+                return '%s[%s]' % (qualname, args)
+
+        return qualname
+
+    def format_annotation_old(self, annotation):
+        # type: (Any) -> str
+        """format_annotation() for py36 or below"""
         if isinstance(annotation, string_types):
             return annotation  # type: ignore
         if isinstance(annotation, typing.TypeVar):  # type: ignore

--- a/tests/test_util_inspect.py
+++ b/tests/test_util_inspect.py
@@ -215,12 +215,7 @@ def test_Signature_annotations():
 
     # TypeVars and generic types with TypeVars
     sig = inspect.Signature(f2).format_args()
-    if sys.version_info < (3, 7):
-        sig == ('(x: typing.List[T], y: typing.List[T_co], z: T) -> '
-                'typing.List[T_contra]')
-    else:
-        sig == ('(x: typing.List[~T], y: typing.List[+T_co], z: T) -> '
-                'typing.List[-T_contra]')
+    assert sig == '(x: List[T], y: List[T_co], z: T) -> List[T_contra]'
 
     # Union types
     sig = inspect.Signature(f3).format_args()

